docs: run e2e tests (#17)
chaos-mesh/website: docs/development_guides/run_e2e_tests.md

---
id: run_e2e_tests
title: Run E2E Tests
sidebar_label: Run E2E Tests
---

> We build our e2e framework based on [kubetest2](https://github.com/kubernetes-sigs/kubetest2) and [ginkgo](https://onsi.github.io/ginkgo/). There are several ways to run e2e tests.

Reviewer: Why would anyone need this? Please explain what e2e is and in what scenarios would users need this?

## Run e2e tests on your local-machine

You could run full e2e tests on your local machine, just execute `./hack/e2e.sh`. It only needs docker installed on your machine.

> It will takes about 20~30 minutes to finish all the test cases.

Reviewer 2: s/It will takes .../It will take ..../g

## Run e2e tests on an existing cluster

> If you just want run apart of test cases, or want run it on an existing cluster, that's a common requirements during developing e2e test or testing compatible between chaos-mesh and cloud-provided kubernetes cluster.

Reviewer 2: a common requirement

> If you just want run apart of test cases, or want run it on an existing cluster, that's a common requirements during developing e2e test or testing compatible between chaos-mesh and cloud-provided kubernetes cluster.

Reviewer 2: "If you just want run apart of test cases, or want run it on an existing cluster"
It should be changed to "want to run".

It supposes that chaos-mesh is already installed on your cluster, you could follow this [document](https://chaos-mesh.org/docs/user_guides/installation) to install chaos-mesh.

There are several images needed by e2e test, for example:

- pingcap/e2e-helper:latest
- nginx:latest

> Some of them are public docker image like `nginx:latest`, kubernetes will pull it from docker registry, so them could be ignored.

Reviewer 2: s/so them could be ignored./so they could be ignored./g

Others are private or already changed with your changes, so you need manually preparing these images.

You could build e2e-helper docker images by

```shell
DOCKER_REGISTRY="" make image-e2e-helper
```

Then you should load this docker image to all nodes in the cluster. You could do this with

```shell
minikube cache add pingcap/e2e-helper:latest
```

or

```shell
kind load docker-image pingcap/e2e-helper:latest
```

After that, you could execute e2e tests by

```shell
make e2e-build

> ./test/image/e2e/bin/ginkgo ./test/image/e2e/bin/e2e.test

Reviewer 3: Idk if it is just my case but my environment didn't set KUBECONFIG by default, so I need to run `export KUBECONFIG=$HOME/.kube/config` before the test

Author: I will append usage for `KUBECONFIG` to switch their target kubernetes cluster. :D

But for your situation, that's wired, as these [codes](https://github.com/kubernetes/kubernetes/blob/5ed4b76a03b5eddc62939a1569b61532b4a06a72/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L145-L154) say, although KUBECONFIG not set, client still try to load config from `$HOME/.kube/config`

Reviewer 3: Yeah but [e2e_test.go](https://github.com/chaos-mesh/chaos-mesh/blob/master/test/e2e/e2e_test.go#L97-L100) explicitly check the value of environment variable `KUBECONFIG` ☹️ maybe we could remove this check

Author: Oops, you are right, `KUBECONFIG` is required for kubernetes test framework.😅 

If we do not provide a `KUBECONFIG`, the test framework will just try in-cluster configuration.

Reviewer 3: Ah got it! Thanks~

It generally takes about 20~30 minutes to finish all the test cases. Therefore, we do not recommend this method.

## Run e2e tests on an existing cluster

For common requirements such as developing e2e test cases or testing compatibilities between Chaos Mesh and Kubernetes clusters on the cloud, you would normally need to run 22e tests on an existing cluster. Take the following steps:

**Note:** Please make sure that Chaos Mesh is already installed on your local.

1. Build e2e-helper docker images by executing:

```shell
DOCKER_REGISTRY="" make image-e2e-helper
```

2. Load this docker image to all nodes in the cluster by executing:

```shell
minikube cache add pingcap/e2e-helper:latest
```

or

```shell
kind load docker-image pingcap/e2e-helper:latest
```

3. Now you should be able to execute e2e tests by executing:

```shell
make e2e-build
./test/image/e2e/bin/ginkgo ./test/image/e2e/bin/e2e.test
```

> "KUBECONFIG" environment variable must be set before run e2e tests, you could also claim it like `KUBECONFIG=/path/to/your/kube/config./test/image/e2e/bin/ginkgo ./test/image/e2e/bin/e2e.test`

## Run specific part of e2e tests

> After you write new e2e test cases, or make changes on certain module, you do not need run full e2e tests. Our e2e tests are built on [ginkgo](https://onsi.github.io/ginkgo/), we could use `--focus` for running part of e2e tests.

Reviewer: ```suggestion
After you write new e2e test cases or make changes on a certain module, you do not need to run full e2e tests. Our e2e tests are built on [ginkgo](https://onsi.github.io/ginkgo/), so you can use `--focus` for running part of e2e tests. 
```

> All e2e test cases are described like:

Reviewer: ```suggestion
For example, we have our full e2e tests described as below:
```

```go
ginkgo.Describe("[Basic]", func() {
  ginkgo.Context("[PodChaos]", func() {
    ginkgo.Context("[PodFailure]", func() {
      ginkgo.It("[Schedule]", func() {
        // test cases
      })
      ginkgo.It("[Pause]", func() {
        // test cases
      })
    })
  })
  ginkgo.Context("[NetworkChaos]", func() {
    ginkgo.Context("[NetworkPartition]", func() {
      ginkgo.It("[Schedule]", func() {
        // test cases
      })
    })
  })
})
```

> We could use `./test/image/e2e/bin/ginkgo -focus="NetworkPartition" ./test/image/e2e/bin/e2e.test` for only running "NetworkPartition" e2e test case.

Reviewer: ```suggestion
You can use `./test/image/e2e/bin/ginkgo -focus="NetworkPartition" ./test/image/e2e/bin/e2e.test` to run only the "NetworkPartition" e2e test case.
```

---
id: run_e2e_tests
title: Run E2E Tests
sidebar_label: Run E2E Tests
---

## Overview

> When developers contribute codes, here is a check for PR called "E2E tests" before PR could be merged.
> 
> E2E means "endpoint to endpoint", it tests chaos-mesh on a real Kubernetes from start to end, for keeping software's correctness and ensuring end-users experience.
> 
> Contributors could run e2e tests on a PR by commenting `/run-e2e-tests`, but it will take a long time for waiting until it's finished. If you want to run e2e on other specific kubernetes cluster, or you only want to focus on part of e2e tests, this document will help you.
> 
> We build our e2e framework based on [kubetest2](https://github.com/kubernetes-sigs/kubetest2) and [ginkgo](https://onsi.github.io/ginkgo/). There are several ways to run e2e tests.

Reviewer: ```suggestion
To ensure the correctness of code contributions, we use  "endpoint to endpoint" (e2e) testing to test Chaos Mesh on a real Kubernetes cluster. This document walks you through how to run e2e tests for each code contribution that you make. 
```

## Run e2e tests on your local-machine

> You could run full e2e tests on your local machine, just execute `./hack/e2e.sh`. It only needs docker installed on your machine.

Reviewer: when would you want to run full e2e tests?

Author: We usually do not execute full e2e tests on the local machine. 🙈 It just introduces a way to run full e2e tests.

> ## Run e2e tests on your local-machine
> 
> You could run full e2e tests on your local machine, just execute `./hack/e2e.sh`. It only needs docker installed on your machine.

Reviewer: ```suggestion
## Run e2e tests on your local-machine

You could run  full e2e tests on your local machine by executing `./hack/e2e.sh`. It requires docker to be installed locally.
```

> It will take about 20~30 minutes to finish all the test cases.

Reviewer: ```suggestion
It generally takes about 20~30 minutes to finish all the test cases. Therefore, we do not recommend this method.
```

## Run e2e tests on an existing cluster

> If you want to run e2e tests on an existing cluster, that's a common requirement during developing e2e tests or testing compatible between chaos-mesh and cloud-provided kubernetes cluster.

Reviewer: ```suggestion
For common requirements such as developing e2e test cases or testing compatibilities between Chaos Mesh and Kubernetes clusters on the cloud, you would normally need to run 22e tests on an existing cluster. Take the following steps:
```

> If you just want to run apart of test cases, or want to run it on an existing cluster, that's a common requirement during developing e2e tests or testing compatible between chaos-mesh and cloud-provided kubernetes cluster.

Reviewer: > If you just want to run apart of test cases"

Should this refer to the next part?

Author: Fixed.

> It supposes that chaos-mesh is already installed on your cluster, you could follow this [document](https://chaos-mesh.org/docs/user_guides/installation) to install chaos-mesh.

Reviewer: ```suggestion
**Note:** Please make sure that Chaos Mesh is already installed on your local.
```

> There are several images needed by e2e tests, for example:
> 
> - pingcap/e2e-helper:latest
> - nginx:latest
> 
> Some of them are public docker images like `nginx:latest`, kubernetes will pull it from docker registry, so they could be ignored.
> 
> Others are private or already changed with your changes, so you need manually preparing these images.

Reviewer: ```suggestion
1. Get the required images as listed below

    - pingcap/e2e-helper:latest
    - nginx:latest
    ```

Reviewer: > Some of them are public docker images like `nginx:latest`, kubernetes will pull it from docker registry, so they could be ignored. Others are private or already changed with your changes, so you need manually preparing these images.

This kind of info is really confusing and redundant. Please consider removing them. Or you can provide links for the images

> You could build e2e-helper docker images by

Reviewer: ```suggestion
2. Build e2e-helper docker images by executing:
```

```shell
DOCKER_REGISTRY="" make image-e2e-helper
```

> Then you should load this docker image to all nodes in the cluster. You could do this with

Reviewer: ```suggestion
3. Load this docker image to all nodes in the cluster by executing:
:
```

```shell
minikube cache add pingcap/e2e-helper:latest
```

or

```shell
kind load docker-image pingcap/e2e-helper:latest
```

> After that, you could execute e2e tests by

Reviewer: ```suggestion
4. Now you should be able to execute e2e tests by executing:
```

Reviewer: For this kind of instructions,  identify the operations and write them in a sequential manner (1. 2. 3. 4 ).  DONOT write it non-stopping with AND, THEN, AFTER THAT.

And DONOT use "you could/should" because it's a step they WILL DO, not something they choose to do.

```shell
make e2e-build

> ./test/image/e2e/bin/ginkgo ./test/image/e2e/bin/e2e.test

Reviewer 4: Have you misses `--install-chaos-mesh`?

```

> "KUBECONFIG" environment variable must be set before run e2e tests, you could also claim it like `KUBECONFIG=/path/to/your/kube/config./test/image/e2e/bin/ginkgo ./test/image/e2e/bin/e2e.test`

## Run specific part of e2e tests

> Our e2e tests are built on [ginkgo](https://onsi.github.io/ginkgo/), we could use `--focus` for running part of e2e tests.

Reviewer: When we should just run part of the e2e?
